rfc/0004-functions.md

```diff
@@ -0,0 +1,131 @@
+- Feature Name: `functions`
+- Start Date: 2021-12-03
+- RFC PR: [supabase/rfcs#0004](https://github.com/supabase/rfcs/pull/0000)
+
+# Summary
+[summary]: #summary
+
+Outlines various options for implementing Supabase Functions.
+
+# Motivation
+[motivation]: #motivation
+
+Firebase offers [Cloud Functions](https://firebase.google.com/docs/functions/), a serverless framework that lets you automatically run backend code in response to events triggered by Firebase features and HTTPS requests. As a Firebase alternative we aim to provide an offering which is as good, or better. Functions is one of our most requested features.
+
+While we could copy the design of Firebase Functions exactly, we also have an opportunity to provide a better experience, and a more flexible implementation - similar to our choice of PostgreSQL over a NoSQL database. 
+
+This RFC will function more as a Design Document - exploring various potential options, and the pros and cons of each.
+
+If you have a use-case for Functions that you can share with the team, please use [this Discussion](https://github.com/supabase/supabase/discussions/4269). It will be incredibly helpful in this Design Doc.
+
+# Roadmap / Blockers
+
+For all of these options, there are a few blockers that we need to clear before proceeding.
+
+- Need to have
+  - API + CLI (in progress). Developers need to authenticate their local machine to deploy their functions to the cloud.
+  - Async Triggers / Function Hooks (in progress) - Ideally we would sign the payload before triggering an external function.
+  - Secrets management / vault (in progress) - a system to store keys/secrets, used to 
+  - Builders on the cloud (not started). Developers should be able to push code to the cloud to be built before running.
+- Nice to have
+  - Realtime logs (in progress). Developers should be able to debug their Functions on the platform.
+  - Domains (not started) - there may be some cases where developers need their functions on their own domains. 
+  - Git Integration (not started). Developers should be able to associate repositories with Supabase projects, and there should be a "function deployment per-branch".
+
+# Option 1: Don't do Functions (partner instead)
+
+Worth considering in any RFC.
+
+Developer platforms like Vercel and Netlify already provide a phenomenal developer experience around Functions. There has been a lot of investment put into these platforms, and we could work with any or all of these - building tight integrations. This is great for the whole ecosystem.
+
+### Pros 
+
+- We can focus on what we're good at (Databases) and make faster progress on some of the other outstanding features like database branching.
+- Cost. Much cheaper for Supabase :)
+
+### Cons
+
+- Supabase developers need to use third-party tools.
+
+# Option 2: Isolates
```

Comment:
Not sure if this is the right spot to leave a comment, on the RFC, sorry if it isn't!

"Option 2: Isolates" seems to include a few function execution environments that, to my knowledge, aren't isolates running on edge, such as Firebase, Netlify and AWS Lambda functions.

I'll use this comment to argue for a proper edge isolate function environment (à la CloudFlare Workers) to be considered 🙃 

Edge isolate functions are, in my opinion and experience, better than their containerized Node.js alternatives for two main reasons:

- **Speed**: since the edge isolate sits closer to the users and has no cold start
- **Cost**: due to their nature, the infrastructure and developer costs can stay crazy cheap. CloudFlare only bills per execution for their isolate workers, so developers save on the usual memory-time costs of containerized functions. This leads to better DX too as developers can deploy "preview versions" of their app at very little cost

I'll also take a crack at answering some of the cons:

- **Hard to run locally**: I'm not sure what would make it hard to run these locally. Depending on how the API is structured, couldn't it be a thin layer on top of an HTTP server? There is some prior art in https://github.com/cloudflare/miniflare
- **No portability. There seems to be a lot of platform lock in for each of these providers.** granted this is true for special CF features like Durable Objects and KV but I would still say that for most API building use cases, the code would be portable. In fact, there's a kind of de facto standard emerging in that world where CloudFlare Workers, Deno Deploy and Remix are all leaning heavily on the standard [Web fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) to handle _incoming requests_ as well as outgoing requests, which I would argue is probably a fairly future proof standard to use.

While their environment is not Node.js, which can be a little confusing at first, the Supabase ecosystem already integrates very well with these functions given the heavy use of PostgREST (with [slight tweaks](https://github.com/supabase/supabase/tree/master/examples/with-cloudflare-workers)).

Either way, I will happily keep using Supabase for my data but just wanted to give my 2 cents!

Reply:
Thanks for the feeback @alexturpin - I guess the key point here is that we should consider latency and cost, and as a result of these considerations running them on Cloudflare Workers is the most attractive option.

I guess we can rule out "non-edge Isolates" since there is very little to gain from that implementation. Let's leave this convo running for a bit and then I can incorporate all your feedback